### PR TITLE
[NavigationDrawer] Fix jump when perferredContentSize changes

### PR DESCRIPTION
**Pull request by codeman7**
### Context
Currently `MDCBottomDrawerViewController` _jumps_ if the preferredContentSize changes. A client is asking for this _jump_ to not happen but instead the drawer will look the same but have more content. 

#### Conditional checks

1. The check to see if `_contentVCPreferredContentSizeHeightCached` is zero is because on initial layout this value is 0 until we call `cacheLayoutCalculations` which sets this value. 
2. The check to see if `percentageFullScreen > 0.5` is there to make sure we never have an initial drawer factor more than design has specified.

### The problem
`MDCBottomDrawerViewController` _jumps_ if the preferredContentSize changes

### The fix
Set initialDrawerFactor to a custom value so that change doesn't cause a jump.

### Videos
| Before | After |
| - | - |
|before|after|

### CODE REVIEW

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 1]
```diff
@@ -136,6 +136,10 @@ @interface MDCBottomDrawerContainerViewController () <UIScrollViewDelegate>
 // The current bottom drawer state.
 @property(nonatomic) MDCBottomDrawerState drawerState;
 
+// The height of the drawer at time of layout, in relation to the screen height, 0.0 is totally
```
Comment (yarneo) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m]: please let the user know this is in percentage and what the default values are in voiceover, landscape, and default.

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 2]
```diff
@@ -475,12 +484,34 @@ - (void)viewDidDisappear:(BOOL)animated {
...
+  }
   _contentHeaderTopInset = NSNotFound;
   _contentHeightSurplus = NSNotFound;
   _addedContentHeight = NSNotFound;
   [self.view setNeedsLayout];
 }
 
+- (void)resetInitialDrawerFactorWithHeight:(CGFloat)height {
+  if (_contentVCPreferredContentSizeHeightCached == 0) {
+    [self cacheLayoutCalculations];
```
Comment (yarneo) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:509]: in what cases would it be 0, why is this conditional needed?

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 3]
```diff
@@ -475,12 +484,34 @@ - (void)viewDidDisappear:(BOOL)animated {
...
+  }
+  CGFloat totalHeight = self.headerViewController.preferredContentSize.height +
+                        _contentVCPreferredContentSizeHeightCached;
+  CGFloat precentageOfFullScreen = totalHeight / self.presentingViewBounds.size.height;
+  if (precentageOfFullScreen > 0.5) {
+    precentageOfFullScreen = 0.5;
+  }
+  self.initialDrawerFactor = precentageOfFullScreen;
+  CGFloat diff = _contentVCPreferredContentSizeHeightCached - height;
+  [self.scrollView setContentOffset:CGPointMake(0, diff)];
```
Comment (yarneo) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m]: why do we need to update the scroll view content offset? can you put a comment in the code explaining?

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 4]
```diff
@@ -475,12 +484,34 @@ - (void)viewDidDisappear:(BOOL)animated {
 
 - (void)preferredContentSizeDidChangeForChildContentContainer:(id<UIContentContainer>)container {
   [super preferredContentSizeDidChangeForChildContentContainer:container];
+  if ([container isKindOfClass:[UIViewController class]]) {
```
Comment (yarneo) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:495]: is this so we don't issue a change when the header view controller height changes?

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 5]
```diff
@@ -284,7 +289,8 @@ - (BOOL)isMobileLandscape {
 }
 
 - (BOOL)shouldPresentFullScreen {
-  return [self isAccessibilityMode] || [self isMobileLandscape];
+  // TODO: Github issue #5828
+  return [self isAccessibilityMode] || [self isMobileLandscape] || _initialDrawerFactor >= 1;
```
Comment (yarneo) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m]: do we need the _initialDrawerFactor >=1 in this case?

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m (hunk 2 above)
Comment (codeman7) [reply components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:509]: Initially it's 0, the first time we set the preferredContentSize of the contentViewController this get's called and at that point _contentVCPreferredContentSzieHeightCached is 0 so we call `cacheLayoutCalculations` to make sure that value is correct.

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 6]
```diff
@@ -475,12 +484,34 @@ - (void)viewDidDisappear:(BOOL)animated {
...
+    }
+  }
   _contentHeaderTopInset = NSNotFound;
   _contentHeightSurplus = NSNotFound;
   _addedContentHeight = NSNotFound;
   [self.view setNeedsLayout];
 }
 
+- (void)resetInitialDrawerFactorWithHeight:(CGFloat)height {
+  if (_contentVCPreferredContentSizeHeightCached == 0) {
```
Comment (romoore) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m]: Tip: Drive-by comment (saw this in an email).  Please use `MDCCGFloatEqual(_contentVCPreferredContentSizeHeightCached, 0)` to account for rounding errors, etc.

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m (hunk 4 above)
Comment (codeman7) [reply components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:495]: `UIContentContainer` may be a presentationController so we have to make sure it's a `UIViewController` we can remove the condition `containerViewController == self.contentViewController` if we want the same behavior for both the header and content.

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m (hunk 5 above)
Comment (codeman7) [reply components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m]: Removed

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 7]
```diff
@@ -284,6 +295,7 @@ - (BOOL)isMobileLandscape {
 }
 
 - (BOOL)shouldPresentFullScreen {
+  // TODO: Github issue #5828
```
Comment (yarneo) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m]: you can remove this //TODO

**Comment by romoore** [comment]
Is this still a WIP or is it ready for review?

**Comment by codeman7** [comment]
This is ready for review.

**Review by yarneo — APPROVED** [review]
please add tests and also a description and a video for the PR, thanks!
